components/builder-api/src/server/authorize.rs

```diff
@@ -23,56 +24,111 @@ use crate::server::{error::{Error,
                             Result},
                     helpers::req_state};
 
+pub enum RoleAction {
+    Read,
+    Write,
+}
+
+impl fmt::Display for RoleAction {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let value = match *self {
+            RoleAction::Read => "read",
+            RoleAction::Write => "write",
+        };
+        write!(f, "{}", value)
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
 pub fn authorize_session(req: &HttpRequest,
-                         origin_opt: Option<&str>)
+                         origin_opt: Option<&str>,
+                         action: RoleAction,
```

Comment:
So it occurs to me that the action here is actually not very useful. The action is pretty much predetermined by the caller context (e.g. ('repo_file_content') and that's defined to be a read or a write in it's very nature. Really all we need is the min_role; read actions can be represented by a lower min_role (e.g something in a write context requires maintainer but read context requires member.) That might simplify some of the logic.

Reply:
It occurs to me that the action field might have been something I suggested; I certainly hadn't realized how decomposed our contexts were.

Reply:
So the intention here is around future flexibility around tunables and user roles in tandem with what we assume is going to be a middleware for handling authz. Effectively, there is a future state where we might want a user to be able configure resource access based on some arbitrary schema. In which case having the route actions will remain and the minimum role will go away. This just felt like a low impact way of prepping for that future work without obligating a full implementation on the immediate.

Reply:
There's a tradeoff between taking complexity today to support future possible needs; you are paying for something you may never use. In this case, I'd argue to keep things simple:
1) I think it will be easy to add this back in if we need it
2) The read/write formulation isn't the only way to describe things (CRUD comes to mind), and we may need to change it anyways.
and probably most importantly:

3) I found it harder to reason about permissions in this current setup; we have two paths (read/write) which behave differently. Coming at it cold it took me a minute to figure out what was intended. I'm leery of complexity in security related code.

Reply:
I agree with @markan and removed the RoleAction in https://github.com/habitat-sh/builder/pull/1378/commits/e3a581121316bba3160d4d3a3a41b380d0e28462